### ISSUE TITLE
Overcritical vue/valid-template-root rule?

### ISSUE THREAD

**Issue by sebiniemann** [vuejs/eslint-plugin-vue#971]
**Tell us about your environment**	
* **ESLint version:** 6.5.1
* **eslint-plugin-vue version:** 5.2.3
* **Node version:**	10.16.0

**Please show your full configuration:**

```json
{
  "env": {
    "browser": true,
    "es6": true,
    "node": true
  },
  "extends": [
    "eslint:recommended",
    "plugin:vue/recommended"
  ],
  "parserOptions": {
    "ecmaVersion": 2019,
    "sourceType": "module"
  },
  "plugins": [
    "vue"
  ]
}
```

**What did you do?**	

```vue
<template>
  <template>
    Hello World
  </template>
</template>
```
https://mysticatea.github.io/vue-eslint-demo/#eJxVjsEKwjAQRH8l7Lmh4DGIZ+8KXryEZiqBmA2bpFBL/91YEepl4e3bYXahgR3I0LHgmYItON2jUv+k1BkhsLqxBLfpfud3QB1JDchkFpoq+sh64DjW7ONDT3pkadOPZA7d5icbvNO/uBbm0tzakY8OsVz8C9vtF69z+nyakx3QmpKVDGkL5CQArW/g3EXT

My actual use case is something like

```vue
<template>
  <template v-if="isValid">
    True
  </template>
  <template v-else>
    False
  </template>
</template>

<script>
export default {
  data() {
    return {
      isValid: false,
    };
  },
};
</script>
```

where we always use `<template>` in conjunction with `v-for`, `v-if`, ...

https://mysticatea.github.io/vue-eslint-demo/#eJx1kM9KxDAQxl9lmJNCQ8FjXT36Ai576iU0EwnEJEwmxd3Sdzdpu4KIp3y/mW/+ZcEpGsIBT0KfyWuh1zEA/BDMytmXEV2+aO/MiFsa4MyFNmP/bx35TIf7TVf9x/4LKuaJXZKq6StFFjBkdfECS6s0WvTD464BmKRwuBPAsd4Atk3q9vD63N61UlOn/uiPHXLxlHFYcC7Uh6imGGzJLnyoWdnIqh2Nw1O35efWWd1XVRyj1NzaoQuGgry7G23eHc/X1L4zJz1RnZQ0Z+IaoJyYCNdvtNp8aA==

**What did you expect to happen?**
No error, as Vue compiles without problem.

**What actually happened?**

> [vue/valid-template-root] The template root disallows 'template' elements.

**Comment by ota-meshi** [comment]
Thank you for this issue.

Your template will fail to compile on [Vue Template Explorer](https://template-explorer.vuejs.org/#%20%20%3Ctemplate%20v-if%3D%22isValid%22%3E%0A%20%20%20%20True%0A%20%20%3C%2Ftemplate%3E%0A%20%20%3Ctemplate%20v-else%3E%0A%20%20%20%20False%0A%20%20%3C%2Ftemplate%3E).

![image](https://user-images.githubusercontent.com/16508807/67011527-c5be1500-f12a-11e9-838e-7595b006ed6a.png)

I think this rule is working correctly.

**Comment by SebastianNiemann** [comment]
Interesting. I tested it within our build process, which uses the official `vue-template-compiler` package and compiles it without failure:

```js
const vueTemplateCompiler = require('vue-template-compiler');

const { render, staticRenderFns } = vueTemplateCompiler.compile(`
<template v-if="isValid">
  True
</template>
<template v-else>
  False
</template>
`);

console.log(render);
console.log(staticRenderFns);
```

Output: 
```
render:  with(this){return (isValid)?[_v("\n  True\n")]:[_v("\n  False\n")]}
staticRenderFns:  []
```

After your comment, I further investigated this behaviour and saw that `vue-template-compiler` will also export warnings besides `render` and `staticRenderFns`. Using this, I extracted the same output as you:

```
Cannot use <template> as component root element because it may contain multiple nodes.
```

After investigating the compiler's source code, I saw that the warning will always be generated if either `<template>` or `<slot>` is present at the root level. Regardless if it proposes an actual problem. The code above for example is fully functional and works as expected within a Vue project.

It would therefore be nice to have an option for `vue/valid-template-root` to allow `<template>` as root, without having to disable the whole rule altogether. As an enhancement, one could also explicitly test for multiple nodes within the root `<template>`.

**Comment by mitar** [comment]
Why has this been closed? I do not think it has been fixed?

**Comment by 716green** [comment]
@sebiniemann  I'm using the new Vue 3 "Vite" with 
`npx create-vite-app <appName>`
Since vite/vue3 SFC, can we look at this again? I'm getting tired of changing all of my settings as I switch between projects

**Comment by sebiniemann** [comment]
Sorry @mitar I must have missed the notification.

After further discussions over another channel, I closed this issue. While VueJS technically supports if/else templates as root, this (unfortunately) does not seems to be an official feature.

@716green I have not yet read up completely on the new VueJS 3 features. Is this now officially supported in the current VueJS 3 alpha (and probably after that)? Would be great.

**Comment by mitar** [comment]
How is this not an official feature in Vue (before 3)? It was asked through an issue https://github.com/vuejs/vue/issues/3878 and support for it explicitly added https://github.com/vuejs/vue/pull/3887 already in 2016.

**Comment by mitar** [comment]
This is separate from https://github.com/vuejs/vue/issues/4128, which I agree it should not be supported. But `v-if` an `v-else` have only one root element always and it is an explicit Vue feature. Please point to any information which shows that it is not such.

**Comment by sebiniemann** [comment]
Thanks for the clarification @mitar 
It looks like I've created a bit of a misunderstanding here.

The main difference is that our use case only uses templates for such manipulations (mainly due to a migration from other frameworks that use non-HTML expressions for loops and if statements). And since it can be further rewritten to avoid `<template>` in this cases, it is fully understandable for me if @ota-meshi decides not to implement this.

As stated last year, after the feedback from @ota-meshi in this thread, the problems lies not within the if/else part, but with the `<template>` tag.

Technically, everything works fine, especially if there is only one element behind it, as the warning says:

    Cannot use <template> as component root element because it **may** contain multiple nodes.

However, `eslint-plugin-vue` seems to be on the more stricter side of *may*, even if the actual condition (not having multiple nodes) isn't violated.

**Comment by mitar** [comment]
I see. I in fact missed that this issue is about `v-if/v-else` in template root in connection with a `template` tag (inside main `template` tag). To me the surprising thing is that I am getting this with regular tags. So in my case:

```
<template>
  <v-row v-if="!indexData">
    ...
  </v-row>
  <custom-component v-else>
    ...
  </custom-component>
</template>
```

So this is just invalid, no?

**Comment by ota-meshi** [comment]
Hi. Sorry for the late reply.

The `vue/valid-template-root` rule is valid when v-if / v-else is set correctly.
We can check it in the [demo](https://mysticatea.github.io/vue-eslint-demo/#eJxNj7sKwzAMRX/FaI4JdAxpx/5AQ6csJr4pBtc2fkEa8u913AfZJJ0jXbTSZCWooz7i6bSIuIyGsV6qzDJX83kkFe5CKzlSJYwNPqE6bZGONnTA17mKUh+lvv3fp4Z80gjUrZQTWmP5ZM2cgjIPnvlsPd+TqTs1lec9nf/Wubc2FrY1pIyEiTf1QnU/7bC4/Z3gxISS5IQP8GWA4DxA2xtAdlAa).

The aims of `vue/valid-template-root` rule is to catch the error beforehand in the general framework of Vue. 

This rule will report an error as `<template>` in the root will cause an error in [Vue Template Explorer](https://template-explorer.vuejs.org/#%20%20%3Ctemplate%20v-if%3D%22isValid%22%3E%0A%20%20%20%20True%0A%20%20%3C%2Ftemplate%3E%0A%20%20%3Ctemplate%20v-else%3E%0A%20%20%20%20False%0A%20%20%3C%2Ftemplate%3E). This is correct behavior.
At this time, we do not have the afford to support other usages, so we do not plan to add optional support.

If you have any other issues, please open a new issue.

**Comment by mitar** [comment]
Awesome, if this works.
